File: example/customersnode/src/spec.ts
import { handlers, invoker, Invoker, Handler, Codec } from './nanobus';

export interface GetCustomerArgs {
  id: number;
}

export interface Customer {
  id: number;
  firstName: string;
  middleName?: string;
  lastName: string;
  email: string;
  address?: Address;
}

export interface Address {
  line1: string;
  line2?: string;
  city: string;
  state: string;
  zip: string;
}

export interface Outbound {
  saveCustomer(customer: Customer): Promise<void>;
  fetchCustomer(id: number): Promise<Customer>;
  customerCreated(customer: Customer): Promise<void>;
}

export class OutboundImpl implements Outbound {
  private invoker: Invoker;

  constructor(invoker: Invoker) {
    this.invoker = invoker;
  }

  async saveCustomer(customer: Customer): Promise<void> {
    await this.invoker('/customers.v1.Outbound/saveCustomer', customer);
  }

  async fetchCustomer(id: number): Promise<Customer> {
    return this.invoker('/customers.v1.Outbound/fetchCustomer', {
      id: id
    });
  }

  async customerCreated(customer: Customer): Promise<void> {
    await this.invoker('/customers.v1.Outbound/customerCreated', customer);
  }
}

export type createCustomerHandler = (customer: Customer) => Promise<Customer>;
export type getCustomerHandler = (id: number) => Promise<Customer>;


function createCustomerWrapper(codec: Codec, handler?: createCustomerHandler): Handler {
  return (input: ArrayBuffer): Promise<ArrayBuffer> => {
    const payload = codec.decoder(input) as Customer;
    return handler(payload).then(result => codec.encoder(result));
  };
}

function getCustomerWrapper(codec: Codec, handler?: getCustomerHandler): Handler {
  return (input: ArrayBuffer): Promise<ArrayBuffer> => {
    const args = codec.decoder(input) as GetCustomerArgs;
    return handler(args.id).then(result => codec.encoder(result));
  };
}

export interface InboundHanders {
  createCustomer?: createCustomerHandler;
  getCustomer?: getCustomerHandler;
}

export function registerInboundHanders(params: InboundHanders): void {
  if (params.createCustomer) {
    handlers.registerHandler(
        '/customers.v1.Inbound/createCustomer',
        createCustomerWrapper(handlers.codec, params.createCustomer));
  }
  if (params.getCustomer) {
    handlers.registerHandler(
      '/customers.v1.Inbound/getCustomer',
        getCustomerWrapper(handlers.codec, params.getCustomer));
  }
}

export const outbound = new OutboundImpl(invoker);

File: example/customerspython/spec.py
from nanobus import Invoker, invoker, handlers
from typing import Awaitable, Callable, Optional
from dataclasses import dataclass


@dataclass
class GetCustomerArgs:
    id: int = 0


@dataclass
class Address:
    line1: str = ''
    line2: Optional[str] = None
    city: str = ''
    state: str = ''
    zip: str = ''


@dataclass
class Customer:
    id: int = 0
    firstName: str = ''
    middleName: Optional[str] = None
    lastName: str = ''
    email: str = ''
    address: Optional[Address] = None


class Outbound:
    async def save_customer(self, customer: Customer):
        pass

    async def customer_created(self, customer: Customer):
        pass

    async def fetch_customer(self, id: int) -> Customer:
        pass


class OutboundImpl(Outbound):
    def __init__(self, invoker: Invoker):
        self.invoker = invoker

    async def save_customer(self, customer: Customer):
        await self.invoker.invoke('customers.v1.Outbound', 'saveCustomer', customer)

    async def customer_created(self, customer: Customer):
        await self.invoker.invoke('customers.v1.Outbound', 'customerCreated', customer)

    async def fetch_customer(self, id: int) -> Customer:
        args = GetCustomerArgs(id=id)
        return await self.invoker.invoke_with_return('customers.v1.Outbound', 'fetchCustomer', args, Customer)


def registerInboundHandlers(
    create_customer: Callable[[Customer], Awaitable[Customer]] = None,
    get_customer: Callable[[int], Awaitable[Customer]] = None,
):
    codec = handlers.codec

    if create_customer != None:
        async def handler(input: bytes) -> bytes:
            customer: Customer = codec.decode(input, Customer)
            result = await create_customer(customer)
            return codec.encode(result)
        handlers.register_handler(
            'customers.v1.Inbound', 'createCustomer', handler)

    if get_customer != None:
        async def handler(input: bytes) -> bytes:
            args: GetCustomerArgs = codec.decode(input, GetCustomerArgs)
            result = await get_customer(args.id)
            return codec.encode(result)
        handlers.register_handler(
            'customers.v1.Inbound', 'getCustomer', handler)


outbound = OutboundImpl(invoker)
